### ISSUE TITLE
EUID in OrganizationIdentifier attribute

### ISSUE THREAD

**Issue by AdrianMueller-sws** [cabforum/smime#263]
The EUID may be used within the OrganizationIdentifier attribute, using the NTR-scheme. This is especially useful for German Commercial Register numbers as these are only unique on a local level. To Be Done:
- Legal/regulatory foundation
- format of EUID, e.g. ....

At the current stage, an change request for ETSI EN 319 412-1 is being discussed. The change in the S/MIME BR has to be in line with the ETSI standard and implemented practices as well.

**Comment by AdrianMueller-sws** [comment]
I suggest to add the following to the S/MIME BR in order to clarify the usage of European Unique Identifier (EUID) as registration reference in the attribute OrganizationIdentifier (for discussion):

### 1.6.1 Definitions
,,,
**European Unique Identifier (EUID):** The EUID uniquely identifies officially registered organizations, Legal Entities and branch offices within the European Union or the European Economic Area. The EUID consists of a prefix designating one of the registers of Member States (‘domestic registers’) and the registration number assigned by the register to the entity. See Appendix A.1.

The EUID is specified in chapter 9 of the Annex contained in the Implementing Regulation (EU) 2021/1042 laying down rules for the application of Directive (EU) 2017/1132 "relating to certain aspects of company law (codification)".
...

### 7.1.4.2.2 Subject distinguished name fields
...
d. **Certificate Field:** subject:organizationIdentifier (2.5.4.97)
...
**Note 4:** For the NTR Registration Scheme, where the Organization or Legal Entity is registered in Germany, the European Unique Identifier (EUID) SHOULD be used as Registration Reference. Where it is registered in another country within the European Union or the European Economic Area, the EUID MAY be used as Registration Reference. Example: `NTRDE-DER3306.HRB66812` For the specifics see [Appendix A.1](#A.1 organizationIdentifier).

###  A.1 organizationIdentifier 
...
For the NTR Registration Scheme used with a European Unique Identifier (EUID) as Registration Reference according to the Implementing Regulation (EU) 2021/1042, the structure of the EUID SHALL be as follows:

- 2 character ISO 3166-1 country code, equal to the one set folllowing the leading string 'NTR'
- the register identifier, for the (member state) domestic register;
- dot-sign ‘.’ (U+002E);
- the registration number, asssigned by the domestic register to the Legal Entity.
- A check digit SHOULD NOT be used.

Example: `NTRDE-DER3306.HRB66812`, where `DE` is the country code for Germany, `R3306` the register ID for the local district court Cologne and `HRB66812` the locally assigned register number for the Legal Entity.

Please note: In Germany the registration numbers are assigned on a local level, not on a national or state-wise level. Therefore, the EUID provides a means for unique registration references on the national level in Germany.
...

**Comment by defacto64** [comment]
I am basically fine with Adrian's proposal, but I have the following doubt: 

if we say that, for the NTR Registration Scheme, where the Organization or Legal Entity is registered in Germany, an EUID **SHOULD** be used as Registration Reference, then it is just a recommendation. Given that in Germany the registration numbers are assigned on a local level, if the CA does not prepend the register identifier (e.g. DER3306), the resulting Registration Reference is not _unique_ as required (see the definition of **Registration Reference** in the Definitions). We have already seen some issues on Bugzilla related to this matter. 

In short, I think that using an EUID, with the NTR scheme for German organizations, should be a **SHALL**, not a SHOULD.

**Comment by CBonnell** [comment]
>A check digit SHOULD NOT be used.

Other Registration Schemes use checksums/digits, such as LEI. Why do we want to discourage in this case?

**Comment by AdrianMueller-sws** [comment]
> I am basically fine with Adrian's proposal, but I have the following doubt:
> 
> if we say that, for the NTR Registration Scheme, where the Organization or Legal Entity is registered in Germany, an EUID **SHOULD** be used as Registration Reference, then it is just a recommendation. Given that in Germany the registration numbers are assigned on a local level, if the CA does not prepend the register identifier (e.g. DER3306), the resulting Registration Reference is not _unique_ as required (see the definition of **Registration Reference** in the Definitions). We have already seen some issues on Bugzilla related to this matter.
> 
> In short, I think that using an EUID, with the NTR scheme for German organizations, should be a **SHALL**, not a SHOULD.

I fully understand Adriano's concern and have no objection to changing the SHOULD to a SHALL for German NTR schema identifiers.

**Comment by AdrianMueller-sws** [comment]
> > A check digit SHOULD NOT be used.
> 
> Other Registration Schemes use checksums/digits, such as LEI. Why do we want to discourage in this case?

I chose SHOULD NOT for the check digit because:

- The change request for ETSI 319 412-1 does not mention the check digit for EUID.
- I have not seen it for the existing usage of EUID, see e.g. the company entries in https://e-justice.europa.eu/489/EN/business_registers__search_for_a_company_in_the_eu .
- Check digits are a simple and effective means for integrity protection of an identifier, e.g. when entered manually into a system. Within a public key certificate, the integrity is already protected otherwise.

However, I do not insist on a SHOULD NOT for this requirement and it could be changed to a MAY (as it is stated in the Implementing Regulation (EU) 2021/1042).

**Comment by AdrianMueller-sws** [comment]
> > (see the definition of Registration Reference in the Definitions).
> 
> "Registration Reference: An identifier assigned to a Legal Entity.". There is no uniqueness requirement in there.
> 
> > In short, I think that using an EUID, with the NTR scheme for German organizations, should be a **SHALL**, not a SHOULD.
> 
> I want to push back on this. I already have concerns with incorporating an EUID within an NTR, and in my opinion EUID should be a Registration Scheme in and of itself, and it feels like we're abusing the NTR scheme for something that just isn't an NTR, even if it does have a relation. But, I realize that ETSI has pretty much accepted such a practise, and we're aligning with that, as such I won't sit in the way of that.
> 
> But making this a SHALL, I feel has some unintended side effects. We've done some research, and we've noticed, among other issues, that the European e-Justice Portal (EJP) does not always return entities when they do exist in a country. If that is a wider issue, it would mean we could find ourselves in a situation where, if the EJP does not return values, we not be able to issue certificate to an organization.
> 
> We have noticed that handelsregister.de also provides the EUID for _some_ records, but that seems to be a minority of records.

- I myself can live with a SHOULD or a SHALL for the EUID within German NTR-OrganizationIdentifier values and am happy either way.
- Regarding the "register identifiers" (prefix designating the "domestic register" or registration agency) used within the EUID: I strongly suggest to use the official code lists that are available, see e.g. https://www.xrepository.de/details/urn:xoev-de:xjustiz:codeliste:gds.bris-listofregisters. I would not make it dependent on an aggregator like BRIS or handelsregister.de using it.

**Comment by srdavidson** [comment]
I have drafted possible ballot text at https://github.com/srdavidson/smime/blob/EUID_proposal/SBR.md

**Comment by srdavidson** [comment]
Updated following today's discussion. https://github.com/srdavidson/smime/compare/cd7c998984050e61fabddda0cbba2e9b02954360...8150e7ac0cf4a6f0c2afd7ad9df98b12330c00bc

**Comment by pfuentes69** [comment]
Hi guys,
coming late to the party... I'd have some questions... 

@defacto64 said "the resulting Registration Reference is not unique as required (see the definition of Registration Reference in the Definitions)", but I fail to find such requirement, as the definition only states "Registration Reference: An identifier assigned to a Legal Entity."

In fact, the requirements state this for the org identifier: "The Registration Reference SHOULD be unique where the Registration Scheme and jurisdiction provide unique identifiers.", which I interpret as a SHOULD requirement for uniqueness where the registration/incorporating agency is providing such uniqueness for a given jurisdiction. Ergo, if in Germany such unique identifiers aren't provided, the uniqueness wouldn't be a requirement.

Certainly I could be missing something, so I would really appreciate if someone helps me to see my mistake, or provides some clarity to understand the problem that we are solving. And if there's a problem to solve, maybe the BR should be changed to clearly request the uniqueness in the orgID.

Thanks!

**Comment by srdavidson** [comment]
Significant clean up of the proposal is at:

https://github.com/srdavidson/smime/compare/c80922087427b1368cb8991eaad4128ef8fe52c0...88431c24d535ec25562461c7576acffcb4938f5b

This incorporates group feedback from the SMCWG call on 2025-02-12 as well as from Inigo on the final form of the ETSI changes.

I think it represents a significant improvement for this section, and allows expansion should additional EUID-type schemes of significance be identified.

**Comment by srdavidson** [comment]
In general, the text that I have written following the WG meeting yesterday has parity with the ETSI approach that is now undergoing approval.
@defacto64 are you suggesting that:
- we remove the DE example at the subdivision level, and restore the recommendation that Germany SHOULD use the EUID?
- we restore the line "The Registration Reference SHOULD be unique where the Registration Scheme and jurisdiction provide unique identifiers."
- Some combination of the above.

**Comment by defacto64** [comment]
First of all, I would not like this to be a discussion between you and me only, and I would appreciate someone else (hopefully at least one S/MIME certified Consumer) having their say here.

I have already heard during the last WG call that it is not required for the `organizationIdentifier` to be unique. While ETSI does not require it to be so with a SHALL, I believe it SHOULD be unique, and I do not think we are bound by ETSI rules more than we need to be. But in reality, I am not _at all_ suggesting that we should deviate from the ETSI rules.

Now, on the introduction of the EUID I have no doubts whatsoever. I think an encoding like NTRcc-\<EUID\> is reasonable and effective, although not particularly useful for countries other than Germany, as country-level unique company numbers are widely available elsewhere in the EU/EEA.

My perplexity stems from your suggestion that in the case of Germany the non-EUID option is _also_ fine together with a subdivision code (i.e. the "plus" encoding NTRDE+\<SP\>-nnnnnn, where <SP> is the 2-letter code of the State or Province that assigned the registration reference nnnnnn). 

The text you are proposing is indeed aligned with the corresponding ETSI EN 319 412-1 update (proposed by Digicert,  if I am not mistaken) that is undergoing approval. I am fine with that, except that the example provided in CR#3 is (IMO) partly flawed:

> EXAMPLE: "NTRDE+HE-123456" (Germany, Hessen), "NTRUS+CA-123456" (United States, California).

While for US, California, the above example works, it does not for Germany, because AFAIK there is no company register "at Hessen level" (HE) nor at any other Land's level. Consequently, although it may be true that company number 123456 was assigned to some German companies within the Hessen Land, it is wrong to use this encoding in as much as the ETSI text specifies that...

> LEG-5.1.4-08: In the case of the "NTR" legal person identity type reference, if the identifiers are **assigned at the subdivision (state or province) level** and are is not unique at the national level, the organizationIdentifier attribute in the subject field shall contain information using one the following structure in the presented order:

But besides being wrong (since in Germany company numbers are assigned by district courts which are not States nor Provinces), using such an encoding can easily result in an _ambiguous_ organizational identifiers. Let's take for example the following two German companies:
* Verwaltungsgesellschaft Hans Höreth GmbH
* Pflasterbau Dull GmbH

According to https://www.handelsregister.de, they are _both_ registered in Bavaria (BY) with company number HRB1000, and therefore their `organizationIdentifier` would be exactly the same: NTRDE+BY-HRB1000, and this is ugly and undesirable IMO.

So what do I propose? Actually, it might be fine for me to not make any changes to the text you have prepared as long as we all agree that a German district court is not a "subdivision" that assigns company registration numbers and therefore the "plus encoding" should not be used in the case of Germany.

Otherwise, we could keep Note 3 but replace "European countries" with "EU/EEA member countries", which should not cause problems for anyone as unique national company numbers are (AFAIK) available everywhere in the EU/EEA except in Germany where EUIDs can be used instead.

**Comment by pfuentes69** [comment]
My personal opinion here is that:
- The SMBR should be reworded to say that the OrgID SHOULD be unique for a given jurisdiction of incorporation and registration, and CAs SHOULD use the information source that facilitates this uniqueness
- Exceptions or specific stipulations such as the one for Germany aren't appropriate as proposed today because aren't in line of the current requirements (no need for uniqueness)... and if the previous proposed rewording occurs, then these specifications aren't needed anymore

Said that, I also say that I pretty much agree with this last comment, so the uniqueness, applied to the whole  subjectName, probably just gives an straightforward solution... and maybe we are again looking for a "masterly solution for a non-existent problem" ;)

**Comment by pfuentes69** [comment]
> What would a "normal person" expect? You look at the certificate and it has some information. You take the OrgID and do the lookup according to the scheme. And then if the name from the lookup doesn't match the name in the cert, you get worried.

I think you are describing a different problem here... If the OrgID (unique or not) is not assigned to the Org, then this is a validation problem and a mis-issuance. 

> Now, how would a normal person react to an OrgID not being unique? Doesn't the ID in OrgID mean that it -identifies- the org?

"Normal people" look at the certificate details? ;)
Kidding apart, as said above, I still think the OrgID SHOULD be unique and that this requirement is included in the SMBR (which is not today)

**Comment by defacto64** [comment]
Obviously, I am not un-aware of the fact that the certificate Subject contains further information that allows the information conveyed by the `organizationIdentifier` alone to be disambiguated. My point is that....
1) an organizationIdentifier should preferably be unambiguous, especially thinking of unattended applications that might (reasonably) expect it to be unique so to be able to use it a lookup key into their databases;
2) in any case, an orgId of "NTRDE\+**HE**\-123456" would not respect the ETSI encoding rule for "subdivisions", since an Amtsgericht is not a State nor a Province.

**Comment by srdavidson** [comment]
Please let me know if the following changes capture the conversation.  They restore some content that had been lost in editing.
https://github.com/srdavidson/smime/commit/d85ed6f54589728e1925deb415513cbf3b31a4fc

**Comment by defacto64** [comment]
While https://github.com/srdavidson/smime/commit/d85ed6f54589728e1925deb415513cbf3b31a4fc still allows the "plus" encoding for Germany (e.g. "NTRDE+HE-123456"), which I believe to be wrong, it's nevertheless an improvement.

**Comment by defacto64** [comment]
Ultimately, it depends on what we think the orgId is for. 

If people here think that it is just one more Subject attribute among the others (e.g. organizationName, localityName, stateOrProvinceName, etc.) and that - for this very reason - it does not necessarily have to be unique, then we can also ignore the fact that an orgId like "NTRDE+BY-HRB1000" does not in itself identify any Bavarian organization precisely. Consequently, a certificate containing such an orgId should be considered OK. It should be noted that, in such a case, since the "BY" is ineffective as a disambiguation element, it could even be omitted, thus arriving at "NTRDE-HRB1000". If one believes that orgIds do not necessarily have to be unique, then this last orgId should also be considered valid. But I would think twice :)

It is perhaps worth noting that this potential ambiguity only arises with the NTR schema (if used in a sloppy way). In fact, all the other registrations schemes (VAT, PSD, LEI) lead to orgIds that are always unique "by construction". It seems unfortunate to me that this may not always be the case with the NTR scheme :)

**Comment by AdrianMueller-sws** [comment]
I summarize the recent discussion as follows:

While one party (including Adriano and me) believes that the content of the organizationIdentifier attribute should be worldwide unique and remaining gaps in the standardization should be filled, other discussion participants argue that organizationIdentifier is just one attribute of the Subject Distinguished Name (SDN) and that only the uniqueness of the entire SDN matters.

Concerning the latter argument: Yes, SDN uniqueness is mandatory (at least within the customer base of one CA), see [RFC5280 chapter 4.1.2.6  Subject](https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.6): "_The DN MUST be unique for each subject entity certified by the one CA as defined by the issuer field._"

However, SDN uniqueness can be (and is) easily achieved without an organizationIdentifier attribute, also when the certificate is issued to an organization: In almost every country, the official organization name is unique throughout the whole country and the combination of the 'Country' and 'Organization' attributes is unique at the time of certificate issuance.
For exceptions like the US and Canada just the 'StateOrProvince' attribute must be included in addition.

Therefore, I suggest to discuss the topic at a basic level: If the goal of the organizationIdentifier is not uniqueness in itself: What purpose does it serve at all? And why is it mandatory according to the S/MIME BR? It could be an optional attribute and uniqueness of the SDN may be achieved as well by the means described above.

**Comment by barrini** [comment]
Hi Stephen, no, this EORI number has not been brought to ESI. The EN 319 412-1 which is under remote consensus now and is the standard responsible of these options does not include this EORI.
I think the use of this number may be very limited but not sure. It can be asked in the upcoming meeting, ESI#85a in 2 weeks and maybe could be included in a future/newer version.

Regards

**Comment by AdrianMueller-sws** [comment]
Following today's discussion I am happy to repost:
Regarding the "register identifiers" (prefix designating the "domestic register" or registration agency) used within the EUID: official code lists are available, see e.g. https://www.xrepository.de/details/urn:xoev-de:xjustiz:codeliste:gds.bris-listofregisters.
